### PR TITLE
[docs] Add Authoring-script-ports.md

### PR DESCRIPTION
**Pull request by ras0219-msft**
Add documentation covering `vcpkg-port-config.cmake` and how to author script ports

### CODE REVIEW

**Review by PhoebeHui — APPROVED** [review]

**Review by dg0yt — CHANGES_REQUESTED** [review]
It seems that this PR lacks the addition of `docs/maintainers/authoring-script-ports.md`.

File: docs/README.md [hunk 1]
```diff
@@ -32,6 +32,7 @@ Vcpkg helps you manage C and C++ libraries on Windows, Linux and MacOS. This too
 - [Manifest files - vcpkg.json](maintainers/manifest-files.md)
 - [Control files](maintainers/control-files.md)
 - [Portfile functions](maintainers/portfile-functions.md)
+- [Authoring script ports](maintainers/authoring-script-ports.md)
```
Comment (strega-nil-ms) [inline docs/README.md:35]:
```suggestion
- [Authoring Script Ports](maintainers/authoring-script-ports.md)
```

File: docs/maintainers/authoring-script-ports.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
+# Authoring Script Ports
+
+Ports can expose functions for other ports to consume during their build, for example the `vcpkg-cmake` helper port exposes the `vcpkg_cmake_configure()` helper function. Packaging common scripts into a shared helper port makes maintenance easier because all consumers can be updated from a single place. Because the scripts come from a port, they can be versioned and depended upon via all the same mechanisms as any other port.
```
Comment (strega-nil-ms) [inline docs/maintainers/authoring-script-ports.md:3]: Can you split this up into multiple lines? Having a single paragraph all on one line makes this difficult to review...

Also, `s/, for example/: as an example,/`

File: docs/maintainers/authoring-script-ports.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
+# Authoring Script Ports
+
+Ports can expose functions for other ports to consume during their build, for example the `vcpkg-cmake` helper port exposes the `vcpkg_cmake_configure()` helper function. Packaging common scripts into a shared helper port makes maintenance easier because all consumers can be updated from a single place. Because the scripts come from a port, they can be versioned and depended upon via all the same mechanisms as any other port.
+
+Script ports are technically implemented via the `vcpkg-port-config.cmake` extension mechanism. Before invoking the `portfile.cmake` of a port, vcpkg will first import `share/<port>/vcpkg-port-config.cmake` from each direct dependency. If the direct dependency is a host dependency, the import will be performed in the host installed tree (e.g. `${HOST_INSTALLED_DIR}/share/${DEP}/vcpkg-port-config.cmake`).
```
Comment (strega-nil-ms) [inline docs/maintainers/authoring-script-ports.md:5]: * `s/ technically //`
* `s/${DEP}/<port>/`

File: docs/maintainers/authoring-script-ports.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
+# Authoring Script Ports
+
+Ports can expose functions for other ports to consume during their build, for example the `vcpkg-cmake` helper port exposes the `vcpkg_cmake_configure()` helper function. Packaging common scripts into a shared helper port makes maintenance easier because all consumers can be updated from a single place. Because the scripts come from a port, they can be versioned and depended upon via all the same mechanisms as any other port.
+
+Script ports are technically implemented via the `vcpkg-port-config.cmake` extension mechanism. Before invoking the `portfile.cmake` of a port, vcpkg will first import `share/<port>/vcpkg-port-config.cmake` from each direct dependency. If the direct dependency is a host dependency, the import will be performed in the host installed tree (e.g. `${HOST_INSTALLED_DIR}/share/${DEP}/vcpkg-port-config.cmake`).
+
+Note that only direct dependencies are checked. This means that if a script relies on another script transitively, it must explicitly import the relevant config.
```
Comment (strega-nil-ms) [inline docs/maintainers/authoring-script-ports.md:7]: * ``s/direct dependencies/the `vcpkg-port-config.cmake` of direct dependencies/``
* `s/are checked/are included in this way/`
* ``s/relevant config/relevant `vcpkg-port-config.cmake`/``

File: docs/maintainers/authoring-script-ports.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+## Example
+
+```cmake
+# ${CURRENT_PACKAGES_DIR}/share/my-helper/vcpkg-port-config.cmake
+
+# This include guard ensures the file will be loaded only once
+include_guard(GLOBAL)
+
+# This is how you could pull in a transitive dependency
+include(${CMAKE_CURRENT_LIST_DIR}/../my-other-helper/vcpkg-port-config.cmake)
```
Comment (strega-nil-ms) [inline docs/maintainers/authoring-script-ports.md:22]:
```suggestion
include("${CMAKE_CURRENT_LIST_DIR}/../my-other-helper/vcpkg-port-config.cmake")
```

File: docs/users/host-dependencies.md [hunk 6]
```diff
@@ -62,4 +60,4 @@ endif()
 
 Some ports are host-only: script ports and tool ports are common examples.
 In this case, you can use the `"native"` supports expression to describe this.
-This supports expression is true when `TARGET_TRIPLET == HOST_TRIPLET`.
+This supports expression is true when `TARGET_TRIPLET == HOST_TRIPLET` / `VCPKG_CROSSCOMPILING` is true.
```
Comment (strega-nil-ms) [inline docs/users/host-dependencies.md:63]:
```suggestion
This supports expression is true when `VCPKG_CROSSCOMPILING` is false (when `TARGET_TRIPLET == HOST_TRIPLET`).
```

**Review by strega-nil-ms — CHANGES_REQUESTED** [review]
It'd also be good to talk about how to add docs to a script port; adding the port to `docs/regenerate.ps1`.

**Comment by dg0yt** [comment]
> It'd also be good to talk about how to add docs to a script port; adding the port to `docs/regenerate.ps1`.

Actually this concept is controversial. It breaks documentation fixes without CI rebuild.
Related: Discussion at https://github.com/microsoft/vcpkg/pull/21763#discussion_r765392565

**Comment by ras0219-msft** [comment]
I have resolved the comments that were more-or-less directly applied but left open the ones where I chose to format at least part of the suggestion differently.

**Review by PhoebeHui — APPROVED** [review]

File: docs/maintainers/authoring-script-ports.md [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+# This include guard ensures the file will be loaded only once
+include_guard(GLOBAL)
+
+# This is how you could pull in a transitive dependency
+include("${CMAKE_CURRENT_LIST_DIR}/../my-other-helper/vcpkg-port-config.cmake")
+
+# Finally, it is convention to put each public function into a separate file with a matching name
+include("${CMAKE_CURRENT_LIST_DIR}/my-helper-function-01.cmake")
+include("${CMAKE_CURRENT_LIST_DIR}/my-helper-function-02.cmake")
+include("${CMAKE_CURRENT_LIST_DIR}/my-helper-function-03.cmake")
```
Comment (strega-nil-ms) [inline docs/maintainers/authoring-script-ports.md:44]:
```suggestion
include("${CMAKE_CURRENT_LIST_DIR}/my_helper_function_01.cmake")
include("${CMAKE_CURRENT_LIST_DIR}/my_helper_function_02.cmake")
include("${CMAKE_CURRENT_LIST_DIR}/my_helper_function_03.cmake")
```

File: docs/maintainers/authoring-script-ports.md [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+Script-to-script dependencies should not be marked as host. The dependency from
+a target port to a script should be marked host, which means that scripts should
+always already be natively compiling. By making script-to-script dependencies
+`"host": false`, it ensures that one script can depend upon the other being in
+its same install directory.
+
+Ports should never provide a `vcpkg-port-config.cmake` file under a different
+`share/` subdirectory than the current port (`${PORT}`).
+
```
Comment (strega-nil-ms) [inline docs/maintainers/authoring-script-ports.md:29]: You should probably add some information on docs; there's a list in `docs/regenerate.ps1` that one should update when adding a new script port.

**Review by strega-nil-ms — CHANGES_REQUESTED** [review]

File: docs/maintainers/authoring-script-ports.md (hunk 8 above)
Comment (ras0219-msft) [reply docs/maintainers/authoring-script-ports.md:29]: I would prefer to merge this PR as-is; then that content can be easily added separately.

**Review by strega-nil-ms — APPROVED** [review]

**Pull request merged**